File: SpinnerActivity/app/src/main/java/com/puresoftware/spinneractivity/MainActivity.java
package com.puresoftware.spinneractivity;

import androidx.appcompat.app.AppCompatActivity;

import android.os.Bundle;
import android.util.Log;
import android.view.MenuItem;
import android.view.View;
import android.widget.Button;
import android.widget.ImageView;
import android.widget.PopupMenu;
import android.widget.Toast;

public class MainActivity extends AppCompatActivity {

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        ImageView btnSpinner = findViewById(R.id.btn_spinner);

        btnSpinner.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {

                PopupMenu popupMenu = new PopupMenu(getApplicationContext(), view);

                // 팝업 할 xml, 팝업 메뉴 클래스
                getMenuInflater().inflate(R.menu.popup, popupMenu.getMenu());

                popupMenu.setOnMenuItemClickListener(new PopupMenu.OnMenuItemClickListener() {
                    @Override
                    public boolean onMenuItemClick(MenuItem menuItem) {

                        int id = menuItem.getItemId();
                        Log.i("TAG", id + "");

                        if (menuItem.getItemId() == R.id.action_menu1) {
                            Toast.makeText(MainActivity.this, "메뉴 1 클릭", Toast.LENGTH_SHORT).show();
                        }
                        return false;
                    }
                });
                popupMenu.show();
            }
        });

    }
}

File: MVP/app/src/main/java/com/puresoftware/mvp/MainActivity.kt
package com.puresoftware.mvp

import androidx.appcompat.app.AppCompatActivity
import android.os.Bundle
import android.util.Log
import android.widget.Toast
import com.puresoftware.mvp.databinding.ActivityMainBinding

class MainActivity : AppCompatActivity(), MvpLoginView {

    private lateinit var binding: ActivityMainBinding
    private lateinit var loginPresenterImpl: LoginPresenterImpl

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivityMainBinding.inflate(layoutInflater)
        setContentView(binding.root)

        loginPresenterImpl = LoginPresenterImpl(this) // Impl에서 로직점검?

        // 버튼을 누르면 Login 시도
        binding.included.loginBtn.setOnClickListener {
            loginPresenterImpl.login()

            Log.i("TAG", "Main의 LoginPresenterImpl 실행")
            Log.i("TAG", "Main의 MvpLoginView.this Data는 ${this.Log()}")
        }
    }

    override val userName: String
        get() = binding.included.etUserName.text.toString()

    override val password: String?
        get() = binding.included.etPassword.text.toString()

    override fun onLoginResult(isLoginSuccess: Boolean?) {
        if (isLoginSuccess == true) {
            Toast.makeText(this@MainActivity, "$userName Login Successful", Toast.LENGTH_SHORT)
                .show()
        } else {
            Toast.makeText(this@MainActivity, "Login Failed", Toast.LENGTH_SHORT).show()
        }
    }
}